docs/adr/0006-process-for-onboarding-new-tenants.md

```diff
@@ -0,0 +1,299 @@
+
+It will also need to allow us to manage licencing, but we anticipate that this will be via an as-yet nonexistent API, as it will be a central service much like tenancy, used by all other services (either directly or via something like APIM).
+
+### Onboarding
+
+Onboarding is a relatively simple part of the process where we create a new Tenant for the client. We will need to determine how we intend licencing to work and what part, if any, the Management API plays..
+
+### Enrollment
+
+Service enrollment is a more interesting aspect of the process. In order to avoid tightly coupling Marain services to the Management API, we need two things:
+- A means of discovering the available services.
+- A means of requesting that a service enroll a Tenant to use it.
+
+As described in ADR 0005, we are envisaging that each service has a Tenant created for it, under a single parent for all Service Tenants. These tenants can then underpin the discovery mechanism that allows the management API to enumerate services that tenants can be enrolled into. Part of the data attached to each tenant will be the URIs at which the endpoints they expose can be found.
+
+Once we have provided a discovery mechanism, we need to define a way in which a service can be asked to enroll a tenant. The simplest way for this to work is to require all Marain services to have a control plane API that implements a set of standard endpoints. We can provide a standard OpenAPI document that defines the endpoints needed, which we would currently expect to be:
+- Request the configuration that is needed to enroll a tenant to the service
+- Enroll a tenant to the service
+- Unenroll a tenant from the service
+
+The idea being that once the Management API has discovered a service, it can request the information that would be required to enroll a tenant to it. This will be a list of required configuration settings, in a known form that allows them to be easily captured and validated before passing them to the enrollment endpoint.
+
+The enrollment endpoint for a service will do two things:
+- Firstly, it will attach the relevant configuration to the tenant that's being enrolled.
+- Secondly, if the service that's being enrolled in has dependencies on other services, it will create a new sub-tenant of its own Service Tenant that will be used to access those dependent services. This new subtenant will then be enrolled for the dependent service.
+
+Whilst we did not wish to tightly couple the Marain services to the Management API, this restriction does not apply for services that depend upon other services. As such if a service does have a dependency, the location of that service (i.e. its URI) will be part of the configuration for that service, making it easy for the enrollment endpoint of a service to create a new tenant and enroll it for the dependent service.
```

Comment:
I'm not sure that a direct service-to-service API call is the best strategy in a cloud hosted environment. It feels brittle and not very resilient.

Reply:
The point with this is that if the dependency already exists, we are not introducing additional coupling.

So for example, with the Workflow service, which depends upon Operations, Workflow is already making direct service-to-service API calls to Operations to fulfil its own requests. Adding another touchpoint (the call to enrollment) doesn't decrease the resiliance.

Reply:
Also, if we were doing some kind of _accept-into-a-queue-and-process-it mechanism_ (as per @mwadams comment below), then we would normally hide this implementation detail from the caller - the enrollment endpoint would return a 202 and, as you say, use the operations service to manage the long running operation. We'd still end up with the direct service-to-service dependency, right?

Reply:
Or was the concern more along the lines of "what happens when part of this chain of enrollments fails?" Which is definitely an issue that we'll need to consider when we get to implementation.

Reply:
Correct. I think Howard was describing the implementation mechanism rather than the concept, but the confusion came in the suggestion of "broadcast" which is specifically not what is happening.

Reply:
It's not just failure cases, but also high load or unavailability scenarios. With provisioning / enrollment processes, it's about expressing the desired state, but also being able to deal with the asynchronicity of failure much further down the dependency chain.

Reply:
We've abandoned this now anyway for a somewhat different approach that only really involves the envisaged management API and the Tenancy service. We will still need to give careful thought to error handling, recovery, etc, but we no longer have to deal with an arbitrarily long chain of service-to-service dependencies.